README.md
# jinqr

[![npm-package](https://img.shields.io/npm/v/jinqr?color=%230a66dc)][npm]
[![code-style](https://img.shields.io/badge/code%20style-prettier+eslint-%230a66dc)][prettier-eslint]

[prettier-eslint]: http://npm.im/prettier-eslint
[npm]: https://npm.im/jinqr

The simple [jomiel] client for downloading media streams.

`jinqr` is a spiritual successor to (now defunct) [cclive].

[jomiel]: https://github.com/guendto/jomiel
[cclive]: https://github.com/guendto/cclive

![demo: jinqr](./images/jinqr-demo.svg)

## Features

- **Configurable**. Supports YAML and JSON configuration. [XDG Base
  Directory Specification] aware.

- **Pipable**. Supports streaming the media being downloaded to the
  [standard output](#examples).

- **Proxy-aware**. Configurable using [environment
  variables](#http-proxy) (HTTP/HTTPS)

- **Client**. Communicates with [jomiel] and media hosting websites.

- **Flexible**. Many output features are customizable.

- **Simplistic**. Small and simple.

## Installation

```shell
npm install -g jinqr
```

## Usage

```text
Usage: jinqr [options] [URI...]

Network:
  -r, --router-endpoint       Jomiel router endpoint address
                                      [string] [default: "tcp://localhost:5514"]
  -t, --connect-timeout       Time allowed connection to jomiel to take
                                                          [number] [default: 30]
      --http-range            Byte range to download, e.g. 12345-67890  [string]
      --http-user-agent       Identify as <string> to HTTP server       [string]
      --http-connect-timeout  Time allowed connection to HTTP servers to take
                                                          [number] [default: 30]

Output:
  -p, --logger-pattern          Specify the logger pattern format to use
                                            [string] [default: "%r %[%p%] - %m"]
  -o, --output-template         Output filename template to use
                        [string] [default: "{title} ({identifier}).{container}"]
  -W, --overwrite-file          Overwrite existing files               [boolean]
      --progressbar-eta-buffer  Number of updates used to calculate the ETA
                                                         [number] [default: 128]
      --progressbar-format      Customize progress bar layout
  [string] [default: "| {bar} {percentage}% || {received}/{expected} | {eta_form
                                                             atted} | {rate}/s"]
      --progressbar-fps         Maximum update rate        [number] [default: 5]
      --progressbar-size        Length of the progress bar in chars
                                                          [number] [default: 25]
      --progressbar-type        Progressbar type to use
  [string] [choices: "legacy", "shades_classic", "shades_grey", "rect"] [default
                                                                       : "rect"]
  -n, --skip-download           Skip download, show details only       [boolean]
  -T, --spinner-type            Spinner type to use, see also --print-Spinners
                                                      [string] [default: "dots"]
  -s, --stream                  Stream profile to download, see --print-streams
                                also                                    [string]
  -l, --verbosity-level         Define verbosity level
   [string] [choices: "all", "trace", "debug", "info", "warn", "error", "fatal",
                                                "mark", "off"] [default: "info"]

Options:
  -v, --version             Show version number                        [boolean]
  -h, --help                Show help                                  [boolean]
  -P, --print-config-paths  Show configuration file paths and exit     [boolean]
  -D, --print-config        Show configuration and exit                [boolean]
  -N, --print-spinners      Show available spinner names and exit      [boolean]
  -S, --print-streams       Show available streams and exit            [boolean]
      --config-file         Load config from file
```

### Examples

Save the media stream to a file.

```shell
jinqr URI
```

Stream the media being downloaded to stdout.

```shell
jinqr -o - URI | mplayer -
```

## Tips and notes

### Defaults to reading from standard input (stdin)

When reading from the stdin, a hash ('#') can be used for comments.

#### Example

```shell
cat > URIs
# This is a comment line and ignored.
https://foo
https://bar  # also ignored.
https://baz
https://foo  # a duplicate.
https://foo  # another duplicate, both will be ignored by jinqr.
```

Consume it with `jinqr`.

```shell
jinqr < URIs
```

### When the streams have no content-length or content-type

`jinqr` will try to inquire the missing values from the HTTP server with
an HTTP HEAD request.

### Negatable flags

To negate a flag, prepend `--no-` to it.

```shell
jinqr `--no-skip-download` URI
```

### Configuration

#### When you prefer JSON configuration over YAML

`jinqr` understands both formats. See [examples/config](https://github.com/guendto/jinqr/tree/master/examples/config)

#### Merged configuration values

The values are read from different sources in the following order (each
step will replace the existing values):

- default values (if any), set by the command line parser
- values read from the configuration files
- values given as the command line args

#### Define the XDG paths

The XDG configuration paths can be defined by the following environment
variables. See also the complete [XDG Base Directory
Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

##### XDG_CONFIG_DIRS

- $XDG_CONFIG_DIRS defines the preference-ordered set of base
  directories to search for configuration files in addition to the
  $XDG_CONFIG_HOME base directory. The directories in $XDG_CONFIG_DIRS
  should be seperated with a colon `:`.

- If $XDG_CONFIG_DIRS is either not set or empty, a value equal to
  /etc/xdg should be used.

##### XDG_CONFIG_HOME

- $XDG_CONFIG_HOME defines the base directory relative to which
  user-specific configuration files should be stored. If
  $XDG_CONFIG_HOME is either not set or empty, a default equal to
  $HOME/.config should be used.

##### Examples

```shell
XDG_CONFIG_HOME=/tmp/config jinqr --print-config-paths
```

### progressbar-format

| Placeholder          |  description                                |
| -------------------- | ------------------------------------------- |
| `rate`               | transfer rate/s                             |
| `expected`           | expected length                             |
| `received`           | amount transferred so far                   |
| `bar`                | progress bar                                |
| `percentage`         | current progress in percent (0-100)         |
| `total`              | end value                                   |
| `value`              | current value set by last update() call     |
| `eta`                | expected time of accomplishment in seconds  |
| `duration`           | elapsed time in seconds                     |
| `eta_formatted`      | expected time of accomplishment (formatted) |
| `duration_formatted` | elapsed time formatted into appropriate     |

The first three are unique to `jinqr`. Refer to the documentation of
`cli-progress` for the most recent list of the
[placeholders](https://github.com/npkgz/cli-progress#bar-formatting) it
supports.

#### Example

```shell
jinqr --progressbar-format "{bar} | {percentage} | {rate}/s" URI
```

### output-template

| Placeholder        |  description                              |
| ------------------ | ----------------------------------------- |
| `author.channelId` | channel ID                                |
| `author.name`      | author name                               |
| `identifier`       | video identifier                          |
| `container`        | video container                           |
| `title`            | video title                               |
| `quality.profile`  | the video quality profile                 |
| `quality.height`   | video quality height                      |
| `quality.width`    | video quality width                       |
| `date.iso8601`     | download date in the ISO8601 format       |
| `date.locale`      | download date based on the locale setting |

##### Example

```shell
jinqr -o ~/Downloads/{author.name}/{title}.{container} URI
```

#### Standard output

When `output-template` is set to `-` the media being downloaded will be
written to the standard output.

#### Output path is missing

When the output path does not exist, `jinqr` will try to create it.

### HTTP proxy

Refer to the documentation of `global-agent` for the [environment
variables](https://github.com/gajus/global-agent#environment-variables).

## License

`jinqr` is licensed under the [Apache License version 2.0][aplv2].

[aplv2]: https://www.tldrlegal.com/l/apache2

## Acknowledgements

`jinqr` supports downloading HTTP streams only.


CHANGES.md
# Changes

All notable changes to this project will be documented in this file.

The format is based on [Keep a
Changelog](https://keepachangelog.com/en/1.0.0/), and this project
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v0.1.0] - 2021-07-08

- First public release.

[unreleased]: https://github.com/guendto/jinqr/compare/v0.1.0..HEAD
[v0.1.0]: https://github.com/guendto/jinqr/releases/tag/v0.1.0


src/options.js
/*
 * -*- coding: utf-8 -*-
 *
 * jinqr
 *
 * Copyright
 *  2021 Toni Gündoğdu
 *
 *
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * The options module of `jinqr`.
 * @module
 */
import { Presets } from "cli-progress";
import { readFileSync, existsSync } from "fs";
import { load as loadYAML } from "js-yaml";

import log4js from "log4js";
import yargs from "yargs";

import { xdgConfigPaths } from "./xdg.js";

/**
 * The class used to parse the command line args.
 */
export class Options {
  /** @static {string} */
  static #GROUP_NETWORK = "Network:";
  /** @static {string} */
  static #GROUP_OUTPUT = "Output:";
  /** @private */
  #name;

  /**
   * Create a new instance.
   *
   * @arg {string} name - the name of the program
   */
  constructor(name) {
    this.#name = name;
  }

  /**
   * Parse and return the command line options.
   *
   * @returns {object} the `jinqr` configuration options
   */
  parse() {
    const configData = this.#xdgReadConfigFiles();
    const argv = process.argv.slice(2);

    return (
      yargs(argv)
        .scriptName(this.#name)
        .usage("Usage: $0 [options] [URI...]")
        .version("version", "0.1.0")
        .alias("version", "v")
        .help("help")
        .alias("help", "h")
        .options({
          ...this.#generalGroup(),
          ...this.#networkGroup(),
          ...this.#outputGroup()
        })
        // See <https://git.io/JGPol> for "strict"
        .strict()
        // See <https://git.io/JGPou> for ".config([key], [desc], [fn])"
        .config("config-file", "Load config from file", path => {
          return this.#readConfigFile(path);
        })
        .config(configData).argv
    );
  }

  // private

  /**
   * Read the configuration files from the XDG base directories.
   *
   * @returns {object} the configuration data
   *
   * @private
   */
  #xdgReadConfigFiles() {
    let result = {};
    const configPaths = xdgConfigPaths(this.#name);
    for (const path of configPaths) {
      if (existsSync(path)) {
        Object.assign(result, this.#readConfigFile(path));
      }
    }
    return result;
  }

  /**
   * Try to read the configuration file.
   *
   * @arg {string} path to the file
   *
   * @returns {object} the configuration data
   *
   * @private
   */
  #readConfigFile = path => loadYAML(readFileSync(path)) || {};

  /**
   * Return the general group of options.
   *
   * @returns {object} the output group
   *
   * @private
   */
  #generalGroup() {
    return {
      "print-config-paths": {
        alias: "P",
        desc: "Show configuration file paths and exit",
        type: "boolean"
      },
      "print-config": {
        alias: "D",
        desc: "Show configuration and exit",
        type: "boolean"
      },
      "print-spinners": {
        alias: "N",
        desc: "Show available spinner names and exit",
        type: "boolean"
      },
      "print-streams": {
        alias: "S",
        desc: "Show available streams and exit",
        type: "boolean"
      }
    };
  }

  /**
   * Return the output group of options.
   *
   * @returns {object} the output group
   *
   * @private
   */
  #outputGroup() {
    return {
      "logger-pattern": {
        alias: "p",
        group: Options.#GROUP_OUTPUT,
        desc: "Specify the logger pattern format to use",
        default: "%r %[%p%] - %m",
        type: "string"
      },
      "output-template": {
        alias: "o",
        group: Options.#GROUP_OUTPUT,
        desc: "Output filename template to use",
        default: "{title} ({identifier}).{container}",
        type: "string"
      },
      "overwrite-file": {
        alias: "W",
        group: Options.#GROUP_OUTPUT,
        desc: "Overwrite existing files",
        type: "boolean"
      },
      "progressbar-eta-buffer": {
        group: Options.#GROUP_OUTPUT,
        desc: "Number of updates used to calculate the ETA",
        default: 128,
        type: "number"
      },
      "progressbar-format": {
        group: Options.#GROUP_OUTPUT,
        desc: "Customize progress bar layout",
        default:
          "| {bar} {percentage}% || {received}/{expected} | {eta_formatted} | {rate}/s",
        type: "string"
      },
      "progressbar-fps": {
        group: Options.#GROUP_OUTPUT,
        desc: "Maximum update rate",
        default: 5,
        type: "number"
      },
      "progressbar-size": {
        group: Options.#GROUP_OUTPUT,
        desc: "Length of the progress bar in chars",
        default: 25,
        type: "number"
      },
      "progressbar-type": {
        group: Options.#GROUP_OUTPUT,
        desc: "Progressbar type to use",
        choices: Object.keys(Presets),
        default: "rect",
        type: "string"
      },
      "skip-download": {
        alias: "n",
        group: Options.#GROUP_OUTPUT,
        desc: "Skip download, show details only",
        type: "boolean"
      },
      "spinner-type": {
        alias: "T",
        group: Options.#GROUP_OUTPUT,
        desc: "Spinner type to use, see also --print-Spinners",
        default: "dots",
        type: "string"
      },
      stream: {
        alias: "s",
        group: Options.#GROUP_OUTPUT,
        desc: "Stream profile to download, see --print-streams also",
        type: "string"
      },
      "verbosity-level": {
        alias: "l",
        group: Options.#GROUP_OUTPUT,
        desc: "Define verbosity level",
        // See <https://git.io/JG74B> for levels
        choices: Object.keys(log4js.levels)
          .filter(level => level !== "levels")
          .map(level => level.toLowerCase()),
        default: "info",
        type: "string"
      }
    };
  }

  /**
   * Return the network group of options.
   *
   * @returns {object} the network group
   *
   * @private
   */
  #networkGroup() {
    return {
      "router-endpoint": {
        alias: "r",
        group: Options.#GROUP_NETWORK,
        desc: "Jomiel router endpoint address",
        default: "tcp://localhost:5514",
        type: "string"
      },
      "connect-timeout": {
        alias: "t",
        group: Options.#GROUP_NETWORK,
        desc: "Time allowed connection to jomiel to take",
        default: 30,
        type: "number"
      },
      "http-range": {
        group: Options.#GROUP_NETWORK,
        desc: "Byte range to download, e.g. 12345-67890",
        type: "string"
      },
      "http-user-agent": {
        group: Options.#GROUP_NETWORK,
        desc: "Identify as <string> to HTTP server",
        type: "string"
      },
      "http-connect-timeout": {
        group: Options.#GROUP_NETWORK,
        desc: "Time allowed connection to HTTP servers to take",
        default: 30,
        type: "number"
      }
    };
  }
}

// factory function for Options.
export default name => new Options(name);
